### PR TITLE
[query] remove fs from HailContext

### PR DESCRIPTION
**Pull request by cseed**
Summary of changes:
- Overhaul tmpdir handling.  Remove most of the old code.  Added local_tmpdir to `init`.  tmpdir is the networked tmpdir.  local_tmpdir is the tmpdir used for local files on both the driver and the executors.  Added tmpdir and localTmpdir to ExecuteContext.  ExecuteContext removes tmp files on close.  Tmp file base is now required, try to give good base names.  Tmp file names  are now generated by being sufficiently random.
- Removed fs from HailContext.  This involved threading ctx and fs through lots of code (most of the changes).
- Added ExecuteContext to EmitModuleBuilder and friends.  This is necessary because EmitMethodBuilder gives generated code access to backend, fs, etc. which are carried by the ctx.
- Some IR (mostly readers, but also VEP, which needs to load the VEP configuration to determine its type) have overall parameters that control their behavior (e.g. the VCF reader path) but have to do IO to determine other state (like the matrix type, determined from the VCF header).  This complicates pretty printing, serialization, and equality.  I clarified this.  In particular, I seperate the parameters (see, for example, MatrixVCFReaderParameters) which are specified on creation and used for serialization and equality from other derived state.  IR no longer close over ctx or fs and they don't need to do IO after their intiial construction.
- MatrixSpec has subspecs for the marginal tables, and TableSpec has the global and rows RVD.  These are now loaded on construction, so lowering no longer neesd to do IO.

### CODE REVIEW

**Comment by akotlar** [comment]
I will finish the review today. Have looked once, now will give a more careful second pass.

**Comment by cseed** [comment]
Thanks Alex, appreciated.  I know it's a big change, but it is needed for a bunch of downstream stuff.

File: hail/python/hail/context.py [hunk 1]
```diff
@@ -198,9 +210,9 @@ def init(sc=None, app_name='Hail', master=None, local='local[*]',
         Minimum file block size in MB.
     branching_factor : :obj:`int`
         Branching factor for tree aggregation.
-    tmp_dir : :obj:`str`
-        Temporary directory for Hail files. Must be a network-visible
-        file path.
+    tmp_dir : :obj:`str`, optional
+        Networked temporary directory.  Must be a network-visible file
```
Comment (akotlar) [inline hail/python/hail/context.py:214]: Not blocking but I find this wordier than necessary. "Must be a network-visible directory path. Default to /tmp." is sufficient I think (also: it's not a file path, and we specify directory in the preceding sentence

File: hail/python/hail/fs/hadoop_fs.py [hunk 2]
```diff
@@ -7,43 +7,48 @@
...
 
 class HadoopFS(FS):
+    def __init__(self, utils_package_object, jfs):
+        self._utils_package_object = utils_package_object
+        self._jfs = jfs
+
     def open(self, path: str, mode: str = 'r', buffer_size: int = 8192):
         if 'r' in mode:
-            handle = io.BufferedReader(HadoopReader(path, buffer_size), buffer_size=buffer_size)
+            handle = io.BufferedReader(HadoopReader(self, path, buffer_size), buffer_size=buffer_size)
```
Comment (akotlar) [inline hail/python/hail/fs/hadoop_fs.py:16]: why is self being passed here?

File: hail/python/hail/utils/tutorial.py [hunk 3]
```diff
@@ -27,9 +27,18 @@ def init_temp_dir():
         tmp_dir = new_local_temp_dir()
 
 
-def _mkdir(jhc, path):
-    if not Env.jutils().dirExists(jhc, path):
-        Env.jutils().mkdir(jhc, path)
+def _dir_exists(fs, path):
+    return fs.exists(path) and fs.is_dir(path)
```
Comment (akotlar) [inline hail/python/hail/utils/tutorial.py:31]: fs.exists is unneeded

File: hail/python/hail/utils/tutorial.py [hunk 4]
```diff
@@ -27,9 +27,18 @@ def init_temp_dir():
...
+
+
+def _file_exists(fs, path):
+    return fs.exists(path) and fs.is_file(path)
```
Comment (akotlar) [inline hail/python/hail/utils/tutorial.py:35]: same

File: hail/src/main/scala/is/hail/HailContext.scala [hunk 5]
```diff
@@ -437,87 +426,16 @@ object HailContext {
...
-    fileAndLineCounts(regex, files, maxLines).mapValues(_.map(_.value)).toArray
-
-  def getTemporaryFile(nChar: Int = 10, prefix: Option[String] = None, suffix: Option[String] = None): String =
-    fs.getTemporaryFile(tmpDir, nChar, prefix, suffix)
 
   def readPartitions[T: ClassTag](
+    fs: FS,
     path: String,
-    partFiles: Array[String],
+    partFiles: IndexedSeq[String],
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/HailContext.scala:433]: does this need to change? you pass in as array below; I know Scala performs implicit conversions, so not a big deal

File: hail/src/main/scala/is/hail/expr/ir/MatrixIR.scala [hunk 6]
```diff
@@ -174,68 +151,67 @@ abstract class MatrixHybridReader extends TableReader with MatrixReader {
...
-    val entriesPath = path + "/entries"
-    val colsPath = path + "/cols"
+    val rowsPath = params.path + "/rows"
+    val entriesPath = params.path + "/entries"
+    val colsPath = params.path + "/cols"
 
     if (mr.dropCols) {
       val tt = TableType(mr.typ.rowType, mr.typ.rowKey, mr.typ.globalType)
-      val trdr: TableReader = TableNativeReader(rowsPath, options, rowsSpec)
+      val trdr: TableReader = new TableNativeReader(TableNativeReaderParameters(rowsPath, params.options), spec.rowsSpec)
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/ir/MatrixIR.scala:214]: are we missing a PR? I see TableNativeReader as a case class and no TableNativeReaderParameters

Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/ir/MatrixIR.scala:214]: Never mind, GitHub had collapsed it from search, and I missed it among other files

File: hail/src/main/scala/is/hail/HailContext.scala (hunk 5 above)
Comment (cseed) [reply hail/src/main/scala/is/hail/HailContext.scala:433]: Yes, I changed it for equality.  IndexedSeq implements element-wise equality, while Array implements reference equality.

**Pull request merged**